Library/Homebrew/keg_relocate.rb

```diff
@@ -165,6 +165,57 @@ def replace_text_in_files(relocation, files: nil)
     changed_files
   end
 
+  def relocate_build_prefix(keg, old_prefix, new_prefix, verbose)
+    # Find binaries which match prefix strings.
+    string_matches = Set.new
+    keg.each_unique_file_matching(old_prefix) do |file|
+      string_matches << file
+    end
+
+    # Remove files that are known to be plain text.
+    # Note that this also removes sharballs, which we currently don't patch
+    # because they appear to break after patching.
+    string_matches -= text_files
+
+    # Only keep files with prefix matches which contain null bytes.  We need to do this
+    # because text_files is not intended to identify all types of text files.
+    binary_string_matches = Set.new
+    keg.each_unique_binary_file(string_matches) do |file|
+      binary_string_matches << file
+    end
```

Comment:
I'm not convinced fewer loops is necessarily better here. I think it's worth minimising file reads but not loops. As far as I understand now, each loop is through a subset of the previous one. If there's a way to combine reads more: I'd say go for it but I think combining loops may harm readability.

Reply:
I agree re minimising file reads instead of loops, but I would also say doing all three loops in one instead is actually much better for readability than otherwise.

Reply:
While it is true that each loop iterates over a subset of the previous one, I do kind of feel like that structure is leftover boilerplate from when I was using a much less effective way of identifying binary files.  Now that we have a very simple strategy for doing this, I am inclined to have it look something like this:
```
text_files_list = text_files # store output of text_files since it has to read all files in keg

keg.each_unique_file_matching(old_prefix) do |file|
   next if text_files_list.include?(file) # skip files we know are plain text
   next if !file.binary_file? # skip files which do not contain null bytes

   # binary patching code
end
```
I think this would actually be easier to read than how it's currently written.  I can push a separate commit implementing this to see how it looks and we can decide if the readability has improved or worsened.

Reply:
That's pretty much exactly what I had in mind.

Also, I wonder if we should be doing the caching in `#text_files` with something like
```diff
diff --git a/Library/Homebrew/keg_relocate.rb b/Library/Homebrew/keg_relocate.rb
index ad779064b..fd0332f15 100644
--- a/Library/Homebrew/keg_relocate.rb
+++ b/Library/Homebrew/keg_relocate.rb
@@ -223,8 +223,9 @@ class Keg
   end
 
   def text_files
-    text_files = []
-    return text_files if !which("file") || !which("xargs")
+    @text_files ||= []
+    return @text_files if !which("file") || !which("xargs")
+    return @text_files unless @text_files.blank?
 
     # file has known issues with reading files on other locales. Has
     # been fixed upstream for some time, but a sufficiently new enough
@@ -239,7 +240,7 @@ class Keg
         next true if Metafiles::EXTENSIONS.include?(pn.extname)
 
         if pn.text_executable?
-          text_files << pn
+          @text_files << pn
           next true
         end
         false
@@ -260,11 +261,11 @@ class Keg
         path = Pathname.new(path)
         next unless files.include?(path)
 
-        text_files << path
+        @text_files << path
       end
     end
 
-    text_files
+    @text_files
   end
 
   def libtool_files
```
There's probably a better way to do this, since it would be useful to also cache the fact that `text_files` is genuinely empty, which this doesn't do.

Reply:
Yeah I was starting to think as we were discussing this that `text_files` should be able to cache its output.  Up until now it's only used internally by `replace_text_in_files`, so it's not called repeatedly.  

The difficulty as you said is that if `text_files` is actually empty, it would still end up trying to scan the keg every time you call it, thinking it hadn't been cached yet, when in reality there simply were no text files.  Maybe we need to initialize it to something other than `[]`, but which it would never sensibly have as a value, and return `@text_files` if doesn't have this other value.

Reply:
Actually it's probably enough for the first line to be
```ruby
return @text_files if @textfiles.present?
```

Reply:
Agreed that the version you pitched is easier @danielnachun, I was wrong and @carlocab was right here 😁 

> `return @text_files if @textfiles.present?`

note that this won't cache `nil` or empty arrays.

Reply:
Maybe we need to have another variable `@no_text_files` initialized to `nil` that we set to `true` if the size of `text_files` is 0, and then the first two lines would be 
```
return @text_files if @text_files.present?
return [] if @no_text_files || !which("file") || !which("xargs")
```

Reply:
Nah, that's not necessary. If we want to cache `nil`: 

`return @text_files if defined?(@ text_files)`

If we want to cache an empty array:

`return @text_files if text_files`

If we want to only cache non-empty arrays:

`return @text_files if text_files.present?`

Reply:
Thanks that's super helpful. I believe we only need to be able to cache empty arrays as `text_files` should be `[]` if nothing is found and not `nil`.  I think I have an idea of how to verify `text_files` is caching within a unit test, and will try to have this cover both the empty case and the non-empty case when I open that smaller PR.

Reply:
I was going back over this in preparation to open a PR, and I realized that the only reason we even needed to use `text_files` was to get rid of sharballs.  However, we could just as easily do this with the following code:
```ruby
keg.each_unique_file_matching(old_prefix) do |file|
      next if !file.binary_file? # skip files which do not contain null bytes
      next if file.text_executable? # skip sharballs

   # binary patching code
end
```
Whether we use `text_files` or the above code, we still only iterate through the files in the keg once, and this way is much simpler and more self-contained.

I'm planning to totally refactor `text_files` and `replace_text_in_files` as part of the next major PR for this to update the bottle spec to store the names of all binaries with prefix matches, and to make sure we're doing prefix replacement in all text files, not the just the limited subset currently returned by `text_files`.  As part of that refactoring I think we'll be able to greatly simplify how `replace_text_in_files` and `text_files` will probably be unneeded.

Reply:
Sounds good.